### ISSUE TITLE
Unclear Ground Truth Annotations in Benchmark Tables

### ISSUE THREAD

**Issue by Allaa-boutaleb** [SuDIS-ZJU/nlcTables#4]
Hello,

Upon inspecting the data further, I came across a worrisome number of puzzling ground-truth annotations in the benchmark tables. For example, for the nlcTables-U benchmark, here are a few illustrative examples:

- **Query Table:** `q_table_145_3_6_1.csv`  
  - **NL Query:**  
    > “I am on the lookout for additional unionable tables where Birthdate is subsequent to 1849-07-20 09:10:04.705072 and the tables feature the keyword Florence Gertrude.”  
  - **Ground Truth Positive:** `dl_table_145_3_6_1.csv`  
  - **Ground Truth Negative:** `dl_table_145_3_6_1_n_7.csv`  
  - **Issue:** The “negative” file actually contains entries matching both the date constraint and the keyword “Florence Gertrude,” and should be labeled positive.

- **Query Table:** `q_table_145_3_6_2.csv`  
  - **NL Query:**  
    > “I am on the lookout for additional unionable tables where Bapt-date is subsequent to 1870-08-10 09:30:31.241292 and the tables feature the keyword Duncan.”  
  - **Ground Truth Positive:** `dl_table_145_3_6_2.csv`  
  - **Ground Truth Negative:** `dl_table_145_3_6_2_n_6.csv`  
  - **Issue:** `dl_table_145_3_6_2_n_6.csv` contains “Duncan” in the Father column (line 24) and a Bapt-date of 1896-02-23 (line 28), which satisfies both conditions and should be marked positive.

This recurs across many query/response pairs. Not just these two. Could it be that negative pairs might contain some rows that satisfy the conditions but not enough? If so, could you please report the thresholds you've used for different conditions in the benchmark generation process? 

---

Another thing, some tables exhibit malformed or nonsensical headers. Here's an example:

```csv
q_table_1_2_3_1.csv:
Commercial St Opp Brookside Rd,Washington St & Broad St,Quincy Center Bus Station
06:57 PM,06:56 PM,07:13 PM
05:40 PM,05:39 PM,05:56 PM
04:09 PM,04:07 PM,04:26 PM
```

---

I also attempted to reproduce the results reported in the paper (e.g., Starmie and Santos models), but have not been able to match the published performance. I’m uncertain whether this discrepancy arises because:

1. The data currently available differs from what was used in your experiments,  
2. Model training parameters have changed, or  
3. Other factors such as sampling/augmentation strategies.

Could you please clarify the exact training parameters used for Starmie in your experiments, including:

- Sampling and augmentation strategies  
- Number of epochs  
- Learning rate schedule  

Thank you for your attention to these issues.

**Comment by Allaa-boutaleb** [comment]
Hello, I've come across another annotation issue: 

**Ambiguous “around {average}” wording used in both larger‑than and smaller‑than splitters**

In **`union_col.py`**, both `split_numerical_larger_table` and `split_numerical_smaller_table` reuse the same `query_average_templates`. As a result, queries like:

> “I want to find more unionable tables, and these tables have **Price around 80**.”

could refer to **either**  
- tables where ≥ 80 % of rows satisfy `Price > segnum`, or  
- tables where ≥ 80 % of rows satisfy `Price ≤ segnum`,  

with no way to tell which logic was applied.

---

**Excerpt from `split_numerical_larger_table`:**
```python
query_larger_templates = [
    "I want to find more unionable tables, and these tables have {caption} larger than {seg_num}.",
    ...
]

query_average_templates = [
    "I want to find more unionable tables, and these tables have {caption} around {template_average}",
    ...
]

# …

if random.choice([True, False]):
    # “larger than” wording
    selected_templates = query_larger_templates[:template_num]
    nl_query = random.choice(selected_templates).format(
        caption=titles[col], seg_num=segnum
    )
else:
    # “around average” wording
    selected_templates = query_average_templates[:template_num]
    nl_query = random.choice(selected_templates).format(
        caption=titles[col], template_average=average
    )

**Comment by Ciciliaclx** [comment]
Hello,

Thank you for your detailed and thoughtful inquiry. I appreciate the time and effort you have put into examining the benchmark tables and providing such insightful feedback. Let me address your concerns one by one:

**1.Annotation Issues in Negative Pairs:**
You are correct that negative pairs may contain some rows that satisfy the conditions but not enough to qualify as positive examples. In our implementation, we ensure that positive examples have a significant majority of rows (e.g., over 80% for numerical) satisfying the conditions, while negative examples have a minority (e.g., less than 50% for numerical). This setting is more in line with real-life scenarios and intentionally raises the bar for the algorithm's ability to discern between examples. This is achieved through the following logic :

```
datalake_below_num = random.randint(0, min(len(below_or_equal_segnum_data), int(datalake_qualifynum / 4)))`
```

This ensures that positive examples have a substantial number of qualifying rows, while negative examples have fewer.

**2.Malformed or Nonsensical Headers:**
We manually annotated the table captions and column names of original tables before splitting to ensure the table semantics. Regarding the q_table_1_2_3_1.csv you've mentioned, these headers actually represent bus stop and the values represent the times when buses arrive at these stops. 

**3.Training Parameters for Starmie:**
For the Starmie model, the training parameters used in our experiments were as follows:
a.Batch size: 64
b.Number of epochs: 3
c.Learning rate: 5e-5

**4.Ambiguous “Around {average}” Wording:**
In split_numerical_larger(small)_table, we calculate the average of the values in the positive examples and use this average to determine the threshold:
```
values = [parse_value(row[col]) for row in datalake_data if parse_value(row[col]) is not None]
if values:  # Ensure the list is not empty
    average = round(np.mean(values))
```

For negative examples, the distribution is different, ensuring that the average is not consistent with the positive examples. By incorporating conditions for greater than, less than, and mean values, we enhance the diversity of the generated examples, thereby more closely approximating the range of actual user queries.

I hope these clarifications help address your concerns. If you have any further questions or need additional details, please do not hesitate to ask. Your feedback is valuable to us, and we are committed to ensuring the accuracy and clarity of our benchmark data.

Best regards,
Lingxi Cui.
